lb: Better loadbalancer API and CLI

tgraf opened the pull request:
- Adds interfaces for service key/value and RevNat key/value
  to hide address family specific code
- Automatically create map files on first update/insert
- Keep structure in host byte order and convert before reading
  or writing from/to map
- Use ParseTCPAddr() to parse IP and port combination
- Allow reading backends from stdin
- Allow specifying multiple backends in update-service command
- Automatically count backends

Signed-off-by: Thomas Graf thomas@cilium.io

aanm commented on bpf/lbmap/lbmap.go:
```diff
@@ -36,26 +41,82 @@ type ServiceKey interface {
+	GetPort() uint16
+
+	// Set the backend index (master: 0, backend: nth backend)
+	SetBackend(int)
```
Why not `SetBackend(uint16)`?

tgraf replied:
I think it's cleaner to hide the actual size behind the API. We might support more than 64K backends at some point.

aanm commented on bpf/lbmap/lbmap.go:
```diff
@@ -36,26 +41,82 @@ type ServiceKey interface {
+	SetPort(uint16)
+
+	// Set reverse NAT identifier
+	SetRevNat(int)
```
Why not `SetRevNat(uint16)`?

aanm commented on bpf/lbmap/lbmap.go:
```diff
@@ -36,26 +41,82 @@ type ServiceKey interface {
+		return svc, nil
+	} else {
+		svc := val.(*Service4Value)
+		svc.Convert()
```
I think you can have this on the same branch:

``` go
var svc ServiceValue
if key.IsIPv6() {
....
}
svc.Convert()
return svc, nil
```

tgraf replied:
Done

aanm commented on bpf/lbmap/ipv6.go:
```diff
@@ -58,12 +58,21 @@ func NewService6Key(ip net.IP, port uint16, slave uint16) *Service6Key {
 
-func (k Service6Key) GetKeyPtr() unsafe.Pointer {
-	return unsafe.Pointer(&k)
+func (k *Service6Key) String() string {
```
`func (k Service6Key) String() string {`

tgraf replied:
Why?

aanm replied:
Does the String() interface works with a pointer to the receiver?

``` go
svc:=Service6Key{}
fmt.Printf("%s", svc)
```

if this works as is then forget my comment

aanm replied:
It doesn't, check comment above about the String() for `RevNat4Value`

aanm commented on cilium/lb/main.go:
```diff
@@ -140,240 +152,242 @@ func init() {
-	} else {
-		fmt.Printf("%v %d %d\n", svcVal.Address, svcVal.Port, svcVal.RevNAT)
-	}
+	keyStr := fmt.Sprintf("%s:%d", svcKey.Address, svcKey.Port)
```
You have the `svcKey.String()`

tgraf replied:
Done

aanm commented on cilium/lb/main.go:
```diff
@@ -140,240 +152,242 @@ func init() {
-	} else {
-		fmt.Printf("%v %d %d\n", svcVal.Address, svcVal.RevNat, svcVal.Port)
-	}
+	keyStr := fmt.Sprintf("[%s]:%d", svcKey.Address, svcKey.Port)
```
You have the `svcKey.String()`

tgraf replied:
Done

aanm commented on cilium/lb/main.go:
```diff
@@ -140,240 +152,242 @@ func init() {
 }
 
 func dumpService4(key bpf.MapKey, value bpf.MapValue) {
 	svcKey := key.(*lbmap.Service4Key)
```
Out of curiosity, this doesn't work?

``` go
svcKey := key.(lbmap.ServiceKey)
```

tgraf replied:
It needs to know the actual type at some point.

aanm commented on cilium/lb/main.go:
```diff
@@ -140,240 +152,242 @@ func init() {
 
-func cliLookupService(ctx *cli.Context) {
-	key := parseServiceKey(ctx, 0)
+	idx := 1
```
`idx := common.FirstFreeServiceID`

tgraf replied:
I don't want to do this unconditionally and make the entire CLI require a consul instance and everything. A better way would be to provide a REST API through the daemon instead.

aanm replied:
I don't think it requires a consul instance, it just makes sure that we are not using different Service IDs, right?

tgraf replied:
OK, makes sense.

aanm reviewed (changes requested):
Will this

aanm commented on bpf/lbmap/ipv4.go:
```diff
@@ -115,43 +147,53 @@ func Service4DumpParser(key []byte, value []byte) (bpf.MapKey, bpf.MapValue, err
+	return &n
+}
+
+func (v *RevNat4Value) String() string {
```
I've checked, remove the pointer, otherwise the printf doesn't work:
with pointer:

``` go
rnv := RevNat4Value{}
fmt.Printf("%s\n", rnv) //Prints something like {%!s(int=2) %!s(int=3)}
```

Without pointer

``` go
rnv := RevNat4Value{}
fmt.Printf("%s\n", rnv) //Prints something the returned value of the String() method
```

tgraf replied:
Weird, it works just fine for me:
`fmt.Printf("%s = %s\n", key, svc)`
results in:
`1 = 4.4.4.4:80`

Is this golang version specific?

aanm replied:
Curious, it works because the key and svc are pointers, that's why my example didn't work.
https://play.golang.org/p/ymuuHZ0Gch

It seems like if the receiver is a pointer then the String method is only used with the variable is a pointer. If the receiver is not a pointer then the String method is used on both cases, whether a variable is a pointer or not.

aanm commented on bpf/lbmap/ipv6.go:
```diff
@@ -143,16 +175,21 @@ type RevNat6Value struct {
-func (k RevNat6Value) GetValuePtr() unsafe.Pointer {
-	return unsafe.Pointer(&k)
+func (k *RevNat6Value) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(k) }
+func (v *RevNat6Value) String() string              { return fmt.Sprintf("%s:%d", v.Address, v.Port) }
```
Pointer here, check comments above

aanm commented on bpf/lbmap/ipv6.go:
```diff
@@ -84,21 +95,40 @@ type Service6Value struct {
 
-func (v *Service6Value) RevNatKey() RevNatKey {
-	return RevNat6Key(v.RevNat)
+func (v *Service6Value) String() string {
```
Pointer here, check comments above

aanm commented on cilium/lb/main.go:
```diff
@@ -123,7 +136,7 @@ func init() {
 					},
 				},
-				ArgsUsage: "--all | <address> <port> <slave-index>",
+				ArgsUsage: "--all | <address>:<port>",
```
is it prepared to deal with the IPv6 `:` + port?

tgraf replied:
What do you mean? `lb delete-service [IPv6]:port` works fine. Do you mean something different?

aanm replied:
ah I see. I think we should keep it as `<address> <port>` (like `ncat`) to avoid confusion whether the user should have entered the `[` and `]` after he realized that `ff00::1:8080` doesn't work. This way it doesn't matter if it's is IPv4 or IPv6. It's my opinion, if you like that way it's fine as well.

tgraf replied:
Let's merge it like this for now. The IP:PORT notation is used for `--frontend` and `--backend` as well. Splitting it will complicate it for those cases.

aanm commented on cilium/lb/main.go:
```diff
@@ -140,241 +153,239 @@ func init() {
+	for k := range backends {
+		key.SetBackend(idx)
+		if err := svc.SetAddress(backends[k].IP); err != nil {
+			// FIXME: Undo the damage that is already done
```
What damage? what happen?

tgraf replied:
If adding a backend fails, all previously added backends of that services will currently remain in the map. They will not be hit because the master service entry with the count is missing though. So they become stale entries.

tgraf committed:
lb: Better loadbalancer API and CLI

 * Adds interfaces for service key/value and RevNat key/value
   to hide address family specific code
 * Automatically create map files on first update/insert
 * Keep structure in host byte order and convert before reading
   or writing from/to map
 * Use ParseTCPAddr() to parse IP and port combination
 * Allow reading backends from stdin
 * Allow specifying multiple backends in update-service command
 * Automatically count backends

Acked-by: Andre Martins <andre@cilium.io>
Acked-by: Madhu Challa <madhu@cilium.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
lb: Treat errors while dumping tables as warnings

This is required to dump an IPv6 table if the IPv4 map has not been created yet.

Acked-by: Andre Martins <andre@cilium.io>
Acked-by: Madhu Challa <madhu@cilium.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>
